stan/math/gpu/copy_submatrix_opencl.hpp

```diff
@@ -0,0 +1,76 @@
+#ifndef STAN_MATH_GPU_COPY_SUBMATRIX_HPP
+#define STAN_MATH_GPU_COPY_SUBMATRIX_HPP
+#ifdef STAN_OPENCL
+#include <stan/math/gpu/matrix_gpu.hpp>
+#include <CL/cl.hpp>
+
+namespace stan {
+namespace math {
+/**
+ * Copies a submatrix of the source matrix to
+ * the destination matrix. The submatrix to copy
+ * starts at (src_offset_rows, src_offset_cols)
+ * and is of size size_rows x size_cols.
+ * The submatrix is copied to the
+ * destination matrix starting at
+ * (dst_offset_rows, dst_offset_cols)
+ *
+ * @param src the source matrix
+ * @param dst the destination submatrix
```

Comment:
Since we’re using the arguments for both in and out, can we:
- update for const correctness? I think src should be const. 
- update doxygen comments to use `[in`, `[out]`, ‘m`[in,out]`
- reorder the arguments so the out parameter is the last argument. 

Questions: 
- does this function require that the output matrix is already the right size? Please document what’s expected. 
- why not return a matrix instead of having a void function?

Reply:
For clairification do you want
```cpp
inline void copy_submatrix(int src_offset_rows, int src_offset_cols,
                           int dst_offset_rows, int dst_offset_cols,
                           int size_rows, int size_cols,
                           matrix_gpu& src, matrix_gpu& dst,) {
```
or
```cpp
inline void copy_submatrix(matrix_gpu& src,
                           int src_offset_rows, int src_offset_cols,
                           int dst_offset_rows, int dst_offset_cols,
                           int size_rows, int size_cols, matrix_gpu& dst) {
```

> (a) does this function require that the output matrix is already the right size? Please document what’s expected.
> (b) why not return a matrix instead of having a void function?

So for this function we are copying a subsection of one matrix to another matrix. For example, with `Foo[1:5, 1:5]` and `Bar[1:10, 1:10]` copying `Foo[1:2, 1:2]` into `Bar[2:4, 2:4]`.

For (a) and the above we do check that the submatrix offsets are within both `Foo` and `Bar`. For (b) we are manipulating a subset of `Bar` and 'returning' that manipulated matrix `Bar`, since we are directly manipulating matrix `Bar` I think it makes sense to not have a return else I fear users will think the matrix `Bar` was not manipulated. IE

```cpp
  matrix_gpu Bar2 = copy_submatrix(Foo, Bar, 1, 1, 0, 0, 2, 2)
```

Reading that I would expect `Bar2` to be a copy of `Bar`, but IRL we actually do manipulate `Bar`.  Does that make sense? If so I'd like to keep it as a void. Unless you have an alt or think my worry is irrational here